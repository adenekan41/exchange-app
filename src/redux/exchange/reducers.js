/* -------------------------- Internal Dependencies ------------------------- */
import ExchangeActions from './types';

export const initialState = {
	isLoading: false,
	from: 'USD',
	to: 'EUR',
	currentRate: 0.8342370902,
	wallet: {
		USD: {
			amount: 400,
		},
		GBP: {
			amount: 500,
		},
		EUR: {
			amount: 700,
		},
	},
};

/**
 * Take Test Reducers
 * @param {Object | Array} state
 * @param {Object} action
 */
const exchangeReducer = (state = initialState, action) => {
	const { type, payload } = action;
	switch (type) {
		case ExchangeActions.GET_CURRENCY:
			return {
				...state,
				isLoading: true,
			};
		case ExchangeActions.GET_CURRENCY_SUCCESS:
			return {
				...state,
				isLoading: false,
				currentRate: payload.data,
			};
		case ExchangeActions.GET_CURRENCY_FAIL:
			return {
				...state,
				isLoading: false,
			};
		case ExchangeActions.UPDATE_FROM_CURRENT_WALLET:
			return {
				...state,
				from: payload,
			};
		case ExchangeActions.UPDATE_TO_CURRENT_WALLET:
			return {
				...state,
				to: payload,
			};
		case ExchangeActions.EXCHANGE_CURRENCY:
			return {
				...state,
				wallet: payload,
			};
		default:
			return { ...state };
	}
};

export default exchangeReducer;
