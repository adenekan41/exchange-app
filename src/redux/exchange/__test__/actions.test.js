/* -------------------------------------------------------------------------- */
/*                            External Dependencies                           */
/* -------------------------------------------------------------------------- */
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import moxios from 'moxios';

/* -------------------------- Internal Dependencies ------------------------- */
import * as exchangeReducers from '../reducers';
import {
	updateFromCurrentWallet,
	updateToCurrentWallet,
	getLatestRates,
	exchangeCurrency,
} from '../actions';
import {
	expectedActions,
	UPDATE_FROM,
	UPDATE_TO,
	GET_LATEST_FAILURE,
	EXCHANGE_CURRENCY,
} from './test-mock';
const createStore = configureMockStore([thunk]);
const store = createStore(exchangeReducers.initialState);
describe('Take Test Action', () => {
	beforeEach(() => {
		moxios.install();
	});
	afterEach(() => moxios.uninstall());

	test('Dispatches the correct action and payload for updating FROM state', () => {
		UPDATE_FROM();

		store.dispatch(updateFromCurrentWallet('USD'));

		expect(store.getActions()).toEqual(expectedActions);
		expect(store.getActions()).toMatchSnapshot();
	});

	test('Dispatches the correct action and payload for updating TO state', () => {
		UPDATE_TO();

		store.dispatch(updateToCurrentWallet('USD'));

		expect(store.getActions()).toEqual(expectedActions);
		expect(store.getActions()).toMatchSnapshot();
	});

	test('Dispatches the correct action and payload for exchanging currency', () => {
		EXCHANGE_CURRENCY();

		store.getState().exchange = {
			wallet: {
				USD: {
					amount: 400,
				},
				GBP: {
					amount: 500,
				},
				EUR: {
					amount: 700,
				},
			},
		};
		store.dispatch(exchangeCurrency({ from: 10, to: 30 }, 'USD', 'EUR'));

		expect(store.getActions()).toEqual(expectedActions);
		expect(store.getActions()).toMatchSnapshot();
	});

	test('Dispatches the correct action and payload for getting latest rates failure', async (done) => {
		moxios.wait(() => {
			const request = moxios.requests.mostRecent();
			request.reject({
				status: 500,
				response: { data: 'question not found' },
			});
		});

		GET_LATEST_FAILURE();
		await store.dispatch(getLatestRates('USD', 'EUR')).then(() => {
			expect(store.getActions()).toEqual(expectedActions);
			expect(store.getActions()).toMatchSnapshot();
		});
		done();
	});
});
