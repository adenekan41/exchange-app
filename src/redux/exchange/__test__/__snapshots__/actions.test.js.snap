// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Take Test Action Dispatches the correct action and payload for exchanging currency 1`] = `
Array [
  Object {
    "payload": "USD",
    "type": "UPDATE_FROM_CURRENT_WALLET",
  },
  Object {
    "payload": "USD",
    "type": "UPDATE_TO_CURRENT_WALLET",
  },
  Object {
    "payload": Object {
      "EUR": Object {
        "amount": 730,
      },
      "GBP": Object {
        "amount": 500,
      },
      "USD": Object {
        "amount": 390,
      },
    },
    "type": "EXCHANGE_CURRENCY",
  },
]
`;

exports[`Take Test Action Dispatches the correct action and payload for getting latest rates failure 1`] = `
Array [
  Object {
    "payload": "USD",
    "type": "UPDATE_FROM_CURRENT_WALLET",
  },
  Object {
    "payload": "USD",
    "type": "UPDATE_TO_CURRENT_WALLET",
  },
  Object {
    "payload": Object {
      "EUR": Object {
        "amount": 730,
      },
      "GBP": Object {
        "amount": 500,
      },
      "USD": Object {
        "amount": 390,
      },
    },
    "type": "EXCHANGE_CURRENCY",
  },
  Object {
    "type": "GET_CURRENCY",
  },
  Object {
    "type": "GET_CURRENCY_FAIL",
  },
]
`;

exports[`Take Test Action Dispatches the correct action and payload for updating FROM state 1`] = `
Array [
  Object {
    "payload": "USD",
    "type": "UPDATE_FROM_CURRENT_WALLET",
  },
]
`;

exports[`Take Test Action Dispatches the correct action and payload for updating TO state 1`] = `
Array [
  Object {
    "payload": "USD",
    "type": "UPDATE_FROM_CURRENT_WALLET",
  },
  Object {
    "payload": "USD",
    "type": "UPDATE_TO_CURRENT_WALLET",
  },
]
`;
